### PR TITLE
move CPI over to CRS and refactor flavorgen

### PR DESCRIPTION
**Pull request by yastij**
Signed-off-by: Yassine TIJANI <ytijani@vmware.com>

**What this PR does / why we need it**: This PR moves our templates to deploying the CPI through CRS and refactors flavorgen for better re-usability across its packages

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
deploy CPI through CRS instead of using the built-in APIs
```

### CODE REVIEW

File: packaging/flavorgen/flavors/crs/cpi.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	cpiObjects = append(cpiObjects, roleBinding)
+
+	cpiService := cloudprovider.CloudControllerManagerService()
+	cpiService.TypeMeta = v1.TypeMeta{
+		Kind:       "Service",
+		APIVersion: corev1.SchemeGroupVersion.String(),
+	}
+	cpiObjects = append(cpiObjects, cpiService)
+
+	cpiDaemonSet := cloudprovider.CloudControllerManagerDaemonSet(cloudprovider.DefaultCPIControllerImage, nil)
```
Comment (yastij) [inline packaging/flavorgen/flavors/crs/cpi.go:87]: we still need to match CPI's version to the kubernetes version

File: packaging/flavorgen/flavors/crs/csi.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	infrav1 "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha3"
+	"sigs.k8s.io/cluster-api-provider-vsphere/packaging/flavorgen/flavors/env"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/services/cloudprovider"
+	"sigs.k8s.io/cluster-api/api/v1alpha3"
+	addonsv1alpha3 "sigs.k8s.io/cluster-api/exp/addons/api/v1alpha3"
+)
+
+// createCrsResourceObjects creates the api objects necessary for CSI to function. Also appends the resources to the CRS
+func CreateCrsResourceObjectsCSI(crs *addonsv1alpha3.ClusterResourceSet, vsphereCluster infrav1.VSphereCluster, cluster v1alpha3.Cluster) []runtime.Object {
```
Comment (yastij) [inline packaging/flavorgen/flavors/crs/csi.go:20]: we still need to remove vsphereCluster usage here

**Review by yastij — COMMENTED** [review]
what still needs to be done, in case someone takes a look

**Comment by gab-satchi** [comment]
e2e tries to apply a `crs-0` which conflicts with the crs being included in the template. https://github.com/yastij/cluster-api-provider-vsphere/pull/2

**Comment by yastij** [comment]
@gab-satchi - can you take a look at the linter issues ?

**Review by yastij — COMMENTED** [review]
/approve

**Comment by gab-satchi** [comment]
just some dead code and gofmt. Sent out a PR: https://github.com/yastij/cluster-api-provider-vsphere/pull/3

**Comment by srm09** [comment]
@yastij Need to resolve conflicts here and this should be ready to merge

**Comment by gab-satchi** [comment]
closing since the copy was merged

/close

**Pull request closed without merging**